build/lib/release.sh

```diff
@@ -466,39 +471,99 @@ function kube::release::package_kube_manifests_tarball() {
   kube::release::create_tarball "${package_name}" "${release_stage}/.."
 }
 
+# Builds tarballs for each test platform containing the appropriate binaries.
+function kube::release::package_test_platform_tarballs() {
+  local platform
+  rm -rf "${RELEASE_STAGE}/test"
+  # KUBE_TEST_SERVER_PLATFORMS is a subset of KUBE_TEST_PLATFORMS,
+  # so process it first.
+  for platform in "${KUBE_TEST_SERVER_PLATFORMS[@]}"; do
+    local platform_tag=${platform/\//-} # Replace a "/" for a "-"
+    local release_stage="${RELEASE_STAGE}/test/${platform_tag}/kubernetes"
+    mkdir -p "${release_stage}/test/bin"
+    cp "${KUBE_TEST_SERVER_BINARIES[@]/#/${LOCAL_OUTPUT_BINPATH}/${platform}/}" \
```

Comment:
What does this `#` mean?

Reply:
Hi @fejta,

I mean, it _looks_ like string manipulation, but dang it's nasty. 

From https://www.tldp.org/LDP/abs/html/arrays.html:
> ```
> string=abcABC123ABCabc
> echo ${string[@]}               # abcABC123ABCabc
> ```

And from http://tldp.org/LDP/abs/html/parameter-substitution.html:

> `${var#Pattern}` Remove from $var the shortest part of $Pattern that matches the front end of $var.

I read it as removing the string evaluated from `/${LOCAL_OUTPUT_BINPATH}/${platform}/` from all of the elements in `${KUBE_TEST_SERVER_BINARIES}`.

I suppose it makes sense if they're trying to use the `cp` command's ability to copy multiple source files into a single destination directory:

```
cp [FLAGS] source_file ... target_directory
```

Maybe trying to remove all but the file name for each element of `${KUBE_TEST_SERVER_BINARIES}` in order to copy them into `"${release_stage}/test/bin/"`?

That's my best guess without spending more time looking at it.

Reply:
Hi @fejta,

From the same [arrays link](https://www.tldp.org/LDP/abs/html/arrays.html) as above:

> ```
> # Substring Removal
> 
> # Removes shortest match from front of string(s).
> 
> echo ${arrayZ[@]#f*r}   # one two three five five
> #               ^       # Applied to all elements of the array.
>                         # Matches "four" and removes it.
> ```

This seems to be the same pattern being used and that I sussed out above. Still, with so many variables in play, I haven't traced out what the actual result is.

Reply:
@akutz was on the right path, but I think this is actually a different operator - `/#`, not `#`.

From the [arrays link](https://www.tldp.org/LDP/abs/html/arrays.html):
```
arrayZ=( one two three four five five )

# Substring Replacement

# Replace front-end occurrences of substring.
echo ${arrayZ[@]/#fi/XY}    # one two three four XYve XYve
#                ^          # Applied to all elements of the array.
```

Because the match string is empty (there's no string after `#` before the next `/`), we end up just prepending the following string. (It's extra confusing because we have `/`s in the "replacement" string.)

It's documented in a few other places it's used in this file, e.g. 
https://github.com/kubernetes/kubernetes/blob/61b676aae1e80673251fbbad3712d0e722b11f0d/build/lib/release.sh#L137-L141

I can add a similar comment for this usage.

(Have I mentioned that the bash side of this PR was far more difficult than the bazel side?)

Reply:
Thank you @ixdy. I didn't realize that `/#` behaved differently. I don't tend to use Bash string manipulation at all, let alone Bash arrays. I usually stick with POSIX-compliant rules and defer to `sed` for replacements.

![maxresdefault](https://user-images.githubusercontent.com/101085/53132974-26169b80-3537-11e9-89d8-ff8ec76d3809.jpg)

Reply:
yeah, I blame @jbeda :) https://github.com/kubernetes/kubernetes/pull/1324

Reply:
definitely a cool trick, though.
https://github.com/kubernetes/kubernetes/pull/1324#discussion_r17577258